### PR TITLE
Fix SqlFederation data type converter NullPointerException

### LINKED ISSUE #30669: SqlFederation data type converter execute NullPointerException
### Which version of ShardingSphere did you use?
mysql version: 5.7.43
master
### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy
### Expected behavior
SQL executes normally
### Actual behavior
NullPointerException
### Reason analyze (If you can)
+ DataLength may be empty.
+ VALUES_MAP key is a DataType in uppercase.

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
`SELECT CAST(1 as char)`
`SELECT CAST('2024-03-27' AS DATE)`
ref: #28895

### PR DESCRIPTION
**Pull request by zihaoAK47**
Fixes #30669.

Changes proposed in this pull request:
  - Fix null pointer exception caused by obtaining DataLength
  - Adjust data type name to uppercase

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/converter/segment/projection/impl/DataTypeConverter.java [hunk 1]
```diff
@@ -45,7 +45,12 @@ public static Optional<SqlNode> convert(final DataTypeSegment segment) {
         if (null == segment) {
             return Optional.empty();
         }
-        return Optional.of(new SqlDataTypeSpec(new SqlBasicTypeNameSpec(Objects.requireNonNull(SqlTypeName.get(segment.getDataTypeName())), segment.getDataLength().getPrecision(), SqlParserPos.ZERO),
-                SqlParserPos.ZERO));
+        return Optional.of(new SqlDataTypeSpec(getSqlBasicTypeNameSpec(segment), SqlParserPos.ZERO));
+    }
+    
+    private static SqlBasicTypeNameSpec getSqlBasicTypeNameSpec(final DataTypeSegment segment) {
```
Comment (strongduanmu) [inline kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/converter/segment/projection/impl/DataTypeConverter.java:51]: Hi @zihaoAK47, can you add e2e test case to cover this logic change?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/DataTypeLengthSegment.java [hunk 2]
```diff
@@ -36,8 +36,9 @@ public final class DataTypeLengthSegment implements SQLSegment {
     private int scale;
     
     /**
-     * get secondNumber.
-     * @return Optional.
+     * Get second number.
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/DataTypeLengthSegment.java:39]: Rename it to Get scale?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/converter/segment/projection/impl/DataTypeConverter.java (hunk 1 above)
Comment (zihaoAK47) [reply kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/converter/segment/projection/impl/DataTypeConverter.java:51]: Hi mentor, the issue arises from executing the cast function in a federated query scenario. I tried to add the following statement to the e2e process:
"select cast(1 as char)"
This SQL statement will result in incorrect column types error in the JDBC scenario, causing the e2e process to fail.
﻿
I tried the following other SQL statements, but the result obtained from executing e2e is different from the function executed on MySQL:
select cast("2024-03-28" as date)
Expected :2024-03-28
Actual   :19810

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/DataTypeLengthSegment.java (hunk 2 above)
Comment (zihaoAK47) [reply parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/DataTypeLengthSegment.java:39]: Format method comments :smile:

**Review by strongduanmu — APPROVED** [review]
LGTM, merged.

**Pull request merged**